File: scala/problem014.scala
object e014 extends App {
  def iter(x: BigInt, n: Int): Int =
    if (hash contains(x)) hash(x) + n
  else if (x == 1) n
  else if (x % 2 == 0) iter(x / 2, n + 1)
    else iter(3 * x + 1, n + 1)

  def choose(a: Pair[Int, Int], b: Pair[Int, Int]) = 
    if (a._2 > b._2) a
  else b

  var hash = scala.collection.mutable.Map[BigInt, Int]()
    val limit = 1000000

  //val l1 = (1 until limit) map (x => (x, iter(x, 1))) sortBy(_._2)
  //println(l1.reverse.head)

  //val l2 = (1 until limit) map (x => (x, iter(x, 1))) reduceLeft{ (a, b) => choose(a, b) }
  //println(l2)

  //lazy val l3: Stream[Int] = 1 #:: (Stream.from(2) map (_ + 1))
  //println(l3.take(20).toList)


  var max = 0
  var m = 0
  for (i <- 1 until limit) {
    m = iter(i, 1)
    hash(i) = m
    if (m > max) {
      println(i, m)
      max = m
    }
  }
}

File: c++/problem014.cc
#include <iostream>

using namespace std;

constexpr int LIMIT = 1000000;
long h[LIMIT];

int iter(long x, int n) {
    if (x < LIMIT && h[x] != 0) return h[x] + n;
    else if (x == 1) return n;
    else if (x % 2 == 0) return iter(x / 2, n + 1);
    else return iter(3 * x + 1, n + 1);
}

int main() {
    int max = 0;
    int m = 0;
    for (int i = 2; i < LIMIT; i++) {
        m = iter(i, 0);
        h[i] = m;
        if (m > max) {
            cout << i << " " << m << endl;
            max = m;
        }
    }
    return 0;
}
